### ISSUE TITLE
Adding ability to change server side encryption key location

### ISSUE THREAD

**Issue by aptalca** [owncloud/core#17401]
Hi,

I think it would be a great idea to allow admins to change the location of encryption keys used in server side encryption, at least the private keys.

Right now, both the private and public keys are stored in the user folders under data. Having the keys in the same location as the secured files obviously creates security issues. I realize that the current implementation is not meant to provide security on the server, but only for shared files on external servers. However, being able to store the keys elsewhere would improve on-server security as well.

It would also allow someone to utilize a NAS for files/data storage (through samba or 9p in kvm).

Even putting all of the users' private key files in one location on the data folder would be fine, because one could at least create a symlink to a secure and encrypted location.

Thanks

**Comment by PVince81** [comment]
@schiesbn what do you think ?

**Comment by aptalca** [comment]
To elaborate further, here's a use case:

I have an unRaid media server (NAS), that is also capable of running VMs (KVM).

I set up an ubuntu VM with full disk encryption, 10GB image file residing on a small SSD. In there, I installed owncloud with nginx, got an SSL certificate verified by StartSSL and pointed my domain to the server. I believe this VM is fairly secure, even with direct access to the disk and image file, the disk encryption would provide security.

As the data location, I selected a user share on the unRaid server. UnRaid takes care of parity protection and backups, and provides lots and lots of storage space. The files are encrypted with server side encryption. The VM can directly access the user share (through 9p in KVM), so the user share is not exported via samba or anything else. Although this seems secure, it would not prevent anyone from accessing all of the data if they have access to the disks. The disks are not encrypted, and although the files are, the keys reside in the same folder as them.

What I did was to move the keys to a location within the secure VM and left symlinks in their location on the NAS. This way, if someone accesses the disks on the NAS, they will only get the encrypted files and not the keys.
